### PR TITLE
Reduce latency for 'executeShell' calls.

### PR DESCRIPTION
**Pull request by ragunathjawahar**

### CODE REVIEW

**Comment by ragunathjawahar** [comment]
1. Since this is a Maven project, IntelliJ would have no trouble configuring the application. Besides,  files inside `.idea` folder usually gets committed differently when several developers work on the same project. **Jake Wharton**, once mentioned that he never opens any open-source projects that contain IDE specific files such as `.idea`. And I frequently use [clean-idea](https://github.com/cesarferreira/clean-idea) to remove any `.idea` folders before I open any open-source project. It just helps us keep commits clean without having to fiddle with the IntelliJ files.

2. @vidstige I didn't profile the code and therefore I don't have any numbers. I was building a toy application that sent input commands to my phone. I noticed a visible lag in the input. When I looked into the source code I noticed that several `OutputStreamWriter` instances were created and weren't properly closed. Therefore, I made it a field and closed it properly after use. This improved the latency in keyboard input to my connected Android device. I use a version of the jar that I built with the proposed changes from my local repo in my application.

**Comment by vidstige** [comment]
1. Using a proper IDE is very important, if you are still using vim or something like that, you're making things harder for yourself. Being able to just open the project and run unit tests is a very important part of jadb. I've tested to just open the pom file (with .idea folder removed), but that did not work, as I could not build anymore without manually fiddling around. If you can fix this use-case, I'd be glad to remove .idea folder. I.e. clean clone, open up in IntelliJ, run unit tests. No other manual steps.
2. The speedup you noticed might as well come from some other source. Please investigate further to understand how your speedup works. Making a fields is unlikely to speed things up. Creating a new instance is very fast in java. Perhaps the closing part is related to speedup? If that is the case, leave it as local variable, but close it when used. In general, I don't want any extra code in jadb, that is not needed in any way.

Also, in any event, you'll need to rebase/merge your pull request due to conflicts.

**Comment by ragunathjawahar** [comment]
1. I cloned the repo and remove the `.idea` folder and also the `*.iml` files. I then used IntelliJ's import feature to import the project as a Maven project and was able to successfully run the test cases. The `src`, `target` and `test` directories were properly configured and I was also able to run the test cases in a single click. I use the community edition and following is the information about my IDE.

````
IntelliJ IDEA 2016.1.3
Build #IC-145.1617, built on June 3, 2016
JRE: 1.8.0_76-release-b198 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
````

Here's a screenshot as well.
idea

**Comment by ragunathjawahar** [comment]
Regarding 2, It is always a good practice to close streams after they are used. Retaining a reference to an existing object is always more performant than creating new instances every time. Jadb currently creates several `OutputStreamWriter` instances and doesn't close them after use (yes, there's GC but `Closeable` implementations should always be closed after use). As a library developer, it is our responsibility to responsibly use resources on applications that depend on our code. There are several discussions on closing streams and not leave it to the GC.

**Comment by vidstige** [comment]
2. Yes, it's very important to close streams and other resources properly, as I wrote in my previous comment. So a fix like that I would merge quickly. As for retaining a reference it's called premature optimization. Look it up. If you want to include it, please back up your "speed up" with numbers. You might be surprised how little you gain.

**Comment by vidstige** [comment]
Oh, and the [close method](https://docs.oracle.com/javase/7/docs/api/java/io/OutputStreamWriter.html#close()) also closes the underlying stream, so remove that close statement to avoid double close.

**Comment by ragunathjawahar** [comment]
The application that I was working on was sending touch input events to a game on my phone via ADB. There was a "noticeable" lag in the events and the game wasn't playable at all. That's when I dug deeper into the code and found this issue with the writer not being closed and new instances of it being created every time a command has to be sent. Now, I am able to play the game as I would if I used my touchscreen.

The delay that I am talking about may be acceptable for most utility apps but not good enough for something like gaming. It's either you "play" or "don't". You can't wait for your character to jump an obstacle 100ms after you tell it to. That's life or death in the game!

To measure performance, I have to deal with two different systems - the app running on my phone which receives events and then the JavaFx application that dispatches those events. Which is not trivial to setup and measure for a hobby project.

There are two fundamental problems with the code in [Transport.java](https://github.com/vidstige/jadb/blob/master/src/se/vidstige/jadb/Transport.java).
  1. Unnecessary object creation.
  2. Resources not being freed up properly.

File: .gitignore [hunk 1]
```diff
@@ -63,3 +61,34 @@ $RECYCLE.BIN/
...
+
+# Unit test / coverage reports
+.coverage
+.tox
+
+#Translations
+*.mo
+
+#Mr Developer
+.mr.developer.cfg
```
Comment (vidstige) [inline .gitignore]: Python files does not feel relevant to ignore, please remove together with Translations files etc. Unit test / coverage reports can stay if you wish.

**Comment by vidstige** [comment]
1. Alright, I tested to open/build as well. I'm ok to merge removing .idea folder aswell. :+1:

**Comment by ragunathjawahar** [comment]
Here's a [post](http://stackoverflow.com/questions/884007/correct-way-to-close-nested-streams-and-writers-in-java/884024#884024) on closing streams, close them individually from outside in. And as per the Javadocs, closing an already closed stream has no effect.

File: .gitignore [hunk 2]
```diff
@@ -1,19 +1,15 @@
 ################
 #   IntelliJ   #
 ################
-/.idea/workspace.xml
-/.idea/tasks.xml
-
+/.idea
```
Comment (vidstige) [inline .gitignore]: please add directories like so `/.idea/`. Note trailing slash.

Comment (ragunathjawahar) [inline .gitignore]: Sorry, my mistake. Fixing it.

**Comment by vidstige** [comment]
As described in the link you posted

> When closing chained streams, you only need to close the outermost stream

Alright, just remove the field or provide measurements on the speed gain you get from not creating a new instance. Perhaps just closing the stream is enough? Realize that when you start a shell, you start a new process on the device, there is networking involved, etc, etc. Creating a new java instance is nothing in this context. This is why I doubt your claim. But perhaps there is something else going on? To profile you don't need to profile on the android side - Actually you can just use a mocked test case, see the unit tests in jadb for more information on how to do this.

To merge this - Either leave it as a local variable, or provide hard numbers on the latency gain.

**Comment by vidstige** [comment]
I'm closing this as there is no sign of progress here. The removal of IDE files is already implemented.

**Pull request closed without merging**